### DISCUSSION TITLE
How to manually release memory of Image after data was loaded?

### DISCUSSION THREAD

**Discussion by alabamagan** [TorchIO-project/torchio#525] [Q&A] [answered]
Suppose I created a list of `LabelImage`, I need to loop them in a covariant fashion:
```python
for a in im_list:
  for b in im_list:
    if a == b: continue
    # Do somthing with a & b
    # need to release memory used by b
...
```
So I only need simultaneously a & b, but after the data was loaded in the inner loop, it persist and there is a waste of memory. Is there a way I can release the memory and revert the `LabelImage` to the stage before the data tensor was loaded?

What I've tried:
```python
del b # in inner loop, doesn't release mem
del b.data # can't delete attribute
```

Cheers

**Comment by fepegar** [comment, marked as answer]
Hi, @alabamagan.

This is probably because instances in `a` and `b` are repeated, so even if you delete them with `del`, you will still be keeping a pointer to them in one of the lists and memory will not be released. A possible solution is iterating over `SubjectsDataset`, which makes *copies* of the subjects. Then, maybe you can use the path of the image or the tensor data to compare two images.

```python
import torchio as tio
subjects = [tio.Subject(im=im) for im in im_list]
dataset = tio.SubjectsDataset(subjects)
for sub1 in dataset:
    for sub2 in dataset:
        if compare(sub1.im, sub2.im):
            do_sth()
```

**Reply by alabamagan** [reply]
Thanks! 
What you proposed seems to work.
